{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "26a05f1a-bc58-40c0-b21a-d762189c5722",
   "metadata": {},
   "source": [
    "Notebook used to download gene expression data from the GDC."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "8a761171-fae3-45d9-a2b4-b5829e29ae1a",
   "metadata": {},
   "outputs": [],
   "source": [
    "import requests\n",
    "import json\n",
    "import re\n",
    "import io\n",
    "import os\n",
    "import sys\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from pathlib import Path\n",
    "\n",
    "datapath = Path(\"../data\")\n",
    "\n",
    "pd.set_option('display.max_rows', 50)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "62908c4c-625e-4daa-8d12-beb358a99eed",
   "metadata": {},
   "source": [
    "# Select primary site\n",
    "\n",
    "Uncomment the desired primary site"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "70ae4a27-3ddf-43ea-b653-9fe33e97e505",
   "metadata": {},
   "outputs": [],
   "source": [
    "# primary_site = \"Kidney\" # 126 replicates\n",
    "# TCGA-KIRC 72 replicates (Adenomas and Adenocarcinomas)\n",
    "# TCGA-KIRP 31 replicates (Adenomas and Adenocarcinomas)\n",
    "# TCGA-KICH 23 replicates (Adenomas and Adenocarcinomas)\n",
    "\n",
    "# primary_site = \"Breast\" # 112 replicates\n",
    "# TCGA-BRCA 111 replicates (Ductal and Lobular Neoplasms)\n",
    "\n",
    "# primary_site = \"Lung\" # 106 replicates\n",
    "# TCGA-LUAD 52 replicates (Adenomas and Adenocarcinomas)\n",
    "# TCGA-LUSC 49 replicates (Squamous Cell Neoplasms)\n",
    "\n",
    "# primary_site = \"Thyroid\" # TCGA-THCA 58 replicates (Adenomas and Adenocarcinomas)\n",
    "\n",
    "# primary_site = \"Prostate\" # 52 replicates\n",
    "# TCGA-PRAD 51 replicates (Adenomas and Adenocarcinomas)\n",
    "\n",
    "# primary_site = \"Colorectal\" # 50 replicates \n",
    "# TCGA-COAD 39 replicates (Adenomas and Adenocarcinomas)\n",
    "\n",
    "primary_site = \"Liver\" # 50 replicates \n",
    "# TCGA-LIHC 50 replicates (Adenomas and Adenocarcinomas)\n",
    "\n",
    "# primary_site = \"Stomach\" # 33 replicates\n",
    "# primary_site = \"Uterus\" # 23 replicates\n",
    "# primary_site = \"Bladder\" # 19 replicates\n",
    "# primary_site = \"Esophagus\" # 8 replicates\n",
    "# primary_site = \"Bile Duct\" # 8 replicates\n",
    "# primary_site = \"Pancreas\" # 4 replicates\n",
    "# primary_site = \"Adrenal Gland\" # 3\n",
    "# primary_site = \"Cervix\" # 3\n",
    "# primary_site = \"Soft Tissue\" # 2\n",
    "# primary_site = \"Thymus\" # 2\n",
    "# primary_site = \"Brain\" # 0\n",
    "# primary_site = \"Ovary\" # 0\n",
    "# primary_site = \"Bone Marrow\" # 0\n",
    "# primary_site = \"Skin\" # 0\n",
    "# primary_site = \"Nervous System\" # 0\n",
    "# primary_site = \"Bone\" # 0\n",
    "# primary_site = \"Eye\" # 0\n",
    "# primary_site = \"Lymph Nodes\" # 0\n",
    "# primary_site = \"Pleura\" # 0\n",
    "# primary_site = \"Testis\" # 0\n",
    "\n",
    "overwrite = False       # Re-download and overwrite existing files if present\n",
    "filterByExpr = True     # Filter low expression genes using edgeR\n",
    "cleanup = True          # Delete downloaded files after merging"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "0a39b569-5d71-4a0a-8284-9b6a2ea758fc",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Storing gene expression data in ../data/liver/liver.csv\n"
     ]
    }
   ],
   "source": [
    "outfolder = Path(datapath, primary_site.lower())\n",
    "!mkdir -p $outfolder\n",
    "\n",
    "outname = primary_site.lower() + \".csv\"\n",
    "outfile = Path(outfolder, outname)\n",
    "print(\"Storing gene expression data in\", outfile)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ee468d2c-afa4-438e-aa2d-be007c37a634",
   "metadata": {},
   "source": [
    "# Fetch IDs for matching N and T samples\n",
    "\n",
    "We start by fetching patient meta data using the GDC API"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "2c3a1219-02aa-406a-af7f-938adcbebf93",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "208035\n"
     ]
    }
   ],
   "source": [
    "fields = [\n",
    "    \"file_name\",\n",
    "    \"cases.submitter_id\",\n",
    "    \"cases.samples.sample_type\",\n",
    "    \"cases.disease_type\",\n",
    "    \"cases.project.project_id\",\n",
    "    \"cases.demographic.gender\",\n",
    "    \"cases.demographic.vital_status\",\n",
    "    \"cases.demographic.days_to_birth\",\n",
    "    \"cases.demographic.days_to_death\",\n",
    "    \"cases.demographic.race\",\n",
    "    \"cases.demographic.ethnicity\",\n",
    "    \"cases.exposures.alcohol_intensity\",\n",
    "    \"cases.exposures.alcohol_history\",\n",
    "    \"cases.exposures.years_smoked\"\n",
    "    ]\n",
    "\n",
    "fields = \",\".join(fields)\n",
    "\n",
    "files_endpt = \"https://api.gdc.cancer.gov/files\"\n",
    "\n",
    "# This set of filters is nested under an 'and' operator.\n",
    "filters = {\n",
    "    \"op\": \"and\",\n",
    "    \"content\":[\n",
    "        {\n",
    "        \"op\": \"in\",\n",
    "        \"content\":{\n",
    "            \"field\": \"cases.project.primary_site\",\n",
    "            \"value\": [primary_site]\n",
    "            }\n",
    "        },\n",
    "        {\n",
    "        \"op\": \"in\",\n",
    "        \"content\":{\n",
    "            \"field\": \"files.experimental_strategy\",\n",
    "            \"value\": [\"RNA-Seq\"]\n",
    "            }\n",
    "        },\n",
    "        {\n",
    "        \"op\": \"in\",\n",
    "        \"content\":{\n",
    "            \"field\": \"files.data_format\",\n",
    "            \"value\": [\"TSV\"]\n",
    "            }\n",
    "        },\n",
    "        {\n",
    "        \"op\": \"in\",\n",
    "        \"content\":{\n",
    "            \"field\": \"files.data_category\",\n",
    "            \"value\": [\"Transcriptome Profiling\"]\n",
    "            }\n",
    "        }\n",
    "    ]\n",
    "}\n",
    "\n",
    "# A POST is used, so the filter parameters can be passed directly as a Dict object.\n",
    "params = {\n",
    "    \"filters\": filters,\n",
    "    \"fields\": fields,\n",
    "    \"format\": \"TSV\",\n",
    "    \"size\": \"200000\"\n",
    "    }\n",
    "\n",
    "# The parameters are passed to 'json' rather than 'params' in this case\n",
    "response = requests.post(files_endpt, headers = {\"Content-Type\": \"application/json\"}, json = params)\n",
    "r = response.content.decode(\"utf-8\")\n",
    "print(len(r))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "2b12ddea-5650-4da8-8059-87ac2c59ad3f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Biological replicates: 50 \n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "Adenomas and Adenocarcinomas    100\n",
       "Name: cases.0.disease_type, dtype: int64"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "TCGA-LIHC    100\n",
       "Name: cases.0.project.project_id, dtype: int64"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "data = io.StringIO(r)\n",
    "df = pd.read_csv(data, sep=\"\\t\")\n",
    "df = df[df[\"file_name\"].str.endswith(\"augmented_star_gene_counts.tsv\")]\n",
    "\n",
    "df_N = df[df[\"cases.0.samples.0.sample_type\"] == \"Solid Tissue Normal\"]\n",
    "df_N = df_N.drop_duplicates(subset=\"cases.0.submitter_id\") # technical replicates\n",
    "\n",
    "df_T = df[df[\"cases.0.samples.0.sample_type\"] == \"Primary Tumor\"]\n",
    "df_T = df_T.drop_duplicates(subset=\"cases.0.submitter_id\") \n",
    "\n",
    "NT = set(df_T[\"cases.0.submitter_id\"]) & set(df_N[\"cases.0.submitter_id\"])\n",
    "print(\"Biological replicates:\", len(NT), \"\\n\")\n",
    "\n",
    "df_N = df_N[df_N[\"cases.0.submitter_id\"].isin(NT)]\n",
    "df_T = df_T[df_T[\"cases.0.submitter_id\"].isin(NT)]\n",
    "df_NT = pd.concat([df_N,df_T])\n",
    "\n",
    "assert(len(df_NT[\"cases.0.submitter_id\"].unique()) == len(df_NT)//2)\n",
    "assert(len(df_NT[df_NT[\"cases.0.samples.0.sample_type\"]==\"Solid Tissue Normal\"])\n",
    "       == len(df_NT[df_NT[\"cases.0.samples.0.sample_type\"]==\"Primary Tumor\"]))\n",
    "\n",
    "display(df_NT[\"cases.0.disease_type\"].value_counts())\n",
    "print(\"\")\n",
    "display(df_NT[\"cases.0.project.project_id\"].value_counts())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "5933b185-4f39-4265-adb5-d28364119ed3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>days_to_birth</th>\n",
       "      <th>days_to_death</th>\n",
       "      <th>ethnicity</th>\n",
       "      <th>gender</th>\n",
       "      <th>race</th>\n",
       "      <th>vital_status</th>\n",
       "      <th>disease_type</th>\n",
       "      <th>alcohol_history</th>\n",
       "      <th>alcohol_intensity</th>\n",
       "      <th>years_smoked</th>\n",
       "      <th>project_id</th>\n",
       "      <th>sample_type</th>\n",
       "      <th>submitter_id</th>\n",
       "      <th>file_name</th>\n",
       "      <th>id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>-26400.0</td>\n",
       "      <td>1135.0</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>female</td>\n",
       "      <td>white</td>\n",
       "      <td>Dead</td>\n",
       "      <td>Adenomas and Adenocarcinomas</td>\n",
       "      <td>Not Reported</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TCGA-LIHC</td>\n",
       "      <td>Primary Tumor</td>\n",
       "      <td>TCGA-BC-A10Q</td>\n",
       "      <td>e576f331-424e-43f0-af27-20e07d91c826.rna_seq.a...</td>\n",
       "      <td>df736f4c-b9d6-4805-bd7d-7bc6d92174d7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>-26400.0</td>\n",
       "      <td>1135.0</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>female</td>\n",
       "      <td>white</td>\n",
       "      <td>Dead</td>\n",
       "      <td>Adenomas and Adenocarcinomas</td>\n",
       "      <td>Not Reported</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TCGA-LIHC</td>\n",
       "      <td>Solid Tissue Normal</td>\n",
       "      <td>TCGA-BC-A10Q</td>\n",
       "      <td>14fc153e-3cb3-4d3f-8259-b73a436f3d0b.rna_seq.a...</td>\n",
       "      <td>6e812349-35e7-44e7-ac7c-a4bc19d81fd0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>-24274.0</td>\n",
       "      <td>308.0</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>female</td>\n",
       "      <td>white</td>\n",
       "      <td>Dead</td>\n",
       "      <td>Adenomas and Adenocarcinomas</td>\n",
       "      <td>Not Reported</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TCGA-LIHC</td>\n",
       "      <td>Primary Tumor</td>\n",
       "      <td>TCGA-BC-A10R</td>\n",
       "      <td>01a67ac4-91db-4a3d-9137-fe2240acc861.rna_seq.a...</td>\n",
       "      <td>d80985cc-be01-452c-931c-65fccbe783d9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>-24274.0</td>\n",
       "      <td>308.0</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>female</td>\n",
       "      <td>white</td>\n",
       "      <td>Dead</td>\n",
       "      <td>Adenomas and Adenocarcinomas</td>\n",
       "      <td>Not Reported</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TCGA-LIHC</td>\n",
       "      <td>Solid Tissue Normal</td>\n",
       "      <td>TCGA-BC-A10R</td>\n",
       "      <td>c0b92413-c9f7-4a67-9ca7-cb065ac9c407.rna_seq.a...</td>\n",
       "      <td>d139256a-c9e8-417b-90af-c466a58d459d</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>-27944.0</td>\n",
       "      <td>837.0</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>male</td>\n",
       "      <td>white</td>\n",
       "      <td>Dead</td>\n",
       "      <td>Adenomas and Adenocarcinomas</td>\n",
       "      <td>Not Reported</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TCGA-LIHC</td>\n",
       "      <td>Primary Tumor</td>\n",
       "      <td>TCGA-BC-A10T</td>\n",
       "      <td>533c5b0b-6eda-46c5-9cf0-8c6decb728c9.rna_seq.a...</td>\n",
       "      <td>35c358f0-4afc-4bd3-a0f4-2741134d6f35</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   days_to_birth  days_to_death               ethnicity  gender   race  \\\n",
       "0       -26400.0         1135.0  not hispanic or latino  female  white   \n",
       "1       -26400.0         1135.0  not hispanic or latino  female  white   \n",
       "2       -24274.0          308.0  not hispanic or latino  female  white   \n",
       "3       -24274.0          308.0  not hispanic or latino  female  white   \n",
       "4       -27944.0          837.0  not hispanic or latino    male  white   \n",
       "\n",
       "  vital_status                  disease_type alcohol_history  \\\n",
       "0         Dead  Adenomas and Adenocarcinomas    Not Reported   \n",
       "1         Dead  Adenomas and Adenocarcinomas    Not Reported   \n",
       "2         Dead  Adenomas and Adenocarcinomas    Not Reported   \n",
       "3         Dead  Adenomas and Adenocarcinomas    Not Reported   \n",
       "4         Dead  Adenomas and Adenocarcinomas    Not Reported   \n",
       "\n",
       "   alcohol_intensity  years_smoked project_id          sample_type  \\\n",
       "0                NaN           NaN  TCGA-LIHC        Primary Tumor   \n",
       "1                NaN           NaN  TCGA-LIHC  Solid Tissue Normal   \n",
       "2                NaN           NaN  TCGA-LIHC        Primary Tumor   \n",
       "3                NaN           NaN  TCGA-LIHC  Solid Tissue Normal   \n",
       "4                NaN           NaN  TCGA-LIHC        Primary Tumor   \n",
       "\n",
       "   submitter_id                                          file_name  \\\n",
       "0  TCGA-BC-A10Q  e576f331-424e-43f0-af27-20e07d91c826.rna_seq.a...   \n",
       "1  TCGA-BC-A10Q  14fc153e-3cb3-4d3f-8259-b73a436f3d0b.rna_seq.a...   \n",
       "2  TCGA-BC-A10R  01a67ac4-91db-4a3d-9137-fe2240acc861.rna_seq.a...   \n",
       "3  TCGA-BC-A10R  c0b92413-c9f7-4a67-9ca7-cb065ac9c407.rna_seq.a...   \n",
       "4  TCGA-BC-A10T  533c5b0b-6eda-46c5-9cf0-8c6decb728c9.rna_seq.a...   \n",
       "\n",
       "                                     id  \n",
       "0  df736f4c-b9d6-4805-bd7d-7bc6d92174d7  \n",
       "1  6e812349-35e7-44e7-ac7c-a4bc19d81fd0  \n",
       "2  d80985cc-be01-452c-931c-65fccbe783d9  \n",
       "3  d139256a-c9e8-417b-90af-c466a58d459d  \n",
       "4  35c358f0-4afc-4bd3-a0f4-2741134d6f35  "
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_NT = df_NT.sort_values(by=[\"cases.0.submitter_id\", \"cases.0.samples.0.sample_type\"]).reset_index(drop=True)\n",
    "ids = list(df_NT[\"id\"])\n",
    "df_NT.columns = [col.split(\".\")[-1] for col in df_NT.columns.to_list()]\n",
    "df_NT.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "28011056-6dae-4a40-9207-686544831ac7",
   "metadata": {},
   "source": [
    "To get matching N and T samples after downloading (which does not preserve order of df_NT)\n",
    "\n",
    "- Sort df_NT by id column (not inplace) and save the new index order in ``sorted_id_indices``\n",
    "- After downloading the files, sort MANIFEST.txt by id\n",
    "- Then replace manifest indices with ``sorted_id_indices`` and sort by index\n",
    "- Matching NT pairs are now grouped together again"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "4f45785b-5be8-4fe6-8daf-71f661c88686",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>days_to_birth</th>\n",
       "      <th>days_to_death</th>\n",
       "      <th>ethnicity</th>\n",
       "      <th>gender</th>\n",
       "      <th>race</th>\n",
       "      <th>vital_status</th>\n",
       "      <th>disease_type</th>\n",
       "      <th>alcohol_history</th>\n",
       "      <th>alcohol_intensity</th>\n",
       "      <th>years_smoked</th>\n",
       "      <th>project_id</th>\n",
       "      <th>sample_type</th>\n",
       "      <th>submitter_id</th>\n",
       "      <th>file_name</th>\n",
       "      <th>id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>85</th>\n",
       "      <td>-26985.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>male</td>\n",
       "      <td>white</td>\n",
       "      <td>Alive</td>\n",
       "      <td>Adenomas and Adenocarcinomas</td>\n",
       "      <td>Not Reported</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TCGA-LIHC</td>\n",
       "      <td>Solid Tissue Normal</td>\n",
       "      <td>TCGA-EP-A3RK</td>\n",
       "      <td>28e28451-9c55-4f01-9ac1-fda75a6f828d.rna_seq.a...</td>\n",
       "      <td>0444bb3d-0844-4333-97b0-e71faca95abb</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55</th>\n",
       "      <td>-26009.0</td>\n",
       "      <td>1005.0</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>female</td>\n",
       "      <td>white</td>\n",
       "      <td>Dead</td>\n",
       "      <td>Adenomas and Adenocarcinomas</td>\n",
       "      <td>Not Reported</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TCGA-LIHC</td>\n",
       "      <td>Solid Tissue Normal</td>\n",
       "      <td>TCGA-DD-A1EJ</td>\n",
       "      <td>c9432a6e-4d79-4de2-8ffa-e3b05129f78f.rna_seq.a...</td>\n",
       "      <td>0940bd97-2818-48f6-a0a5-f955152409b8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>-22968.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>not reported</td>\n",
       "      <td>female</td>\n",
       "      <td>white</td>\n",
       "      <td>Alive</td>\n",
       "      <td>Adenomas and Adenocarcinomas</td>\n",
       "      <td>Not Reported</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TCGA-LIHC</td>\n",
       "      <td>Solid Tissue Normal</td>\n",
       "      <td>TCGA-BC-A216</td>\n",
       "      <td>1432142c-908d-46d7-9a2a-2597a0f00ae8.rna_seq.a...</td>\n",
       "      <td>095b68ff-e317-4667-b1d1-ff69254d146a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>-15410.0</td>\n",
       "      <td>1149.0</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>male</td>\n",
       "      <td>black or african american</td>\n",
       "      <td>Dead</td>\n",
       "      <td>Adenomas and Adenocarcinomas</td>\n",
       "      <td>Not Reported</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TCGA-LIHC</td>\n",
       "      <td>Solid Tissue Normal</td>\n",
       "      <td>TCGA-DD-A114</td>\n",
       "      <td>8e595756-ee6b-4fd9-9844-6001ea97fa2e.rna_seq.a...</td>\n",
       "      <td>0c11ccb8-4405-4a2a-9180-b483fb845968</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>-27944.0</td>\n",
       "      <td>837.0</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>male</td>\n",
       "      <td>white</td>\n",
       "      <td>Dead</td>\n",
       "      <td>Adenomas and Adenocarcinomas</td>\n",
       "      <td>Not Reported</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TCGA-LIHC</td>\n",
       "      <td>Solid Tissue Normal</td>\n",
       "      <td>TCGA-BC-A10T</td>\n",
       "      <td>ebbdae1d-a199-4577-a73a-77e9c6a55a92.rna_seq.a...</td>\n",
       "      <td>0dbea4cc-32b6-42af-b4f0-fa0a67b05a7f</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    days_to_birth  days_to_death               ethnicity  gender  \\\n",
       "85       -26985.0            NaN  not hispanic or latino    male   \n",
       "55       -26009.0         1005.0  not hispanic or latino  female   \n",
       "19       -22968.0            NaN            not reported  female   \n",
       "27       -15410.0         1149.0  not hispanic or latino    male   \n",
       "5        -27944.0          837.0  not hispanic or latino    male   \n",
       "\n",
       "                         race vital_status                  disease_type  \\\n",
       "85                      white        Alive  Adenomas and Adenocarcinomas   \n",
       "55                      white         Dead  Adenomas and Adenocarcinomas   \n",
       "19                      white        Alive  Adenomas and Adenocarcinomas   \n",
       "27  black or african american         Dead  Adenomas and Adenocarcinomas   \n",
       "5                       white         Dead  Adenomas and Adenocarcinomas   \n",
       "\n",
       "   alcohol_history  alcohol_intensity  years_smoked project_id  \\\n",
       "85    Not Reported                NaN           NaN  TCGA-LIHC   \n",
       "55    Not Reported                NaN           NaN  TCGA-LIHC   \n",
       "19    Not Reported                NaN           NaN  TCGA-LIHC   \n",
       "27    Not Reported                NaN           NaN  TCGA-LIHC   \n",
       "5     Not Reported                NaN           NaN  TCGA-LIHC   \n",
       "\n",
       "            sample_type  submitter_id  \\\n",
       "85  Solid Tissue Normal  TCGA-EP-A3RK   \n",
       "55  Solid Tissue Normal  TCGA-DD-A1EJ   \n",
       "19  Solid Tissue Normal  TCGA-BC-A216   \n",
       "27  Solid Tissue Normal  TCGA-DD-A114   \n",
       "5   Solid Tissue Normal  TCGA-BC-A10T   \n",
       "\n",
       "                                            file_name  \\\n",
       "85  28e28451-9c55-4f01-9ac1-fda75a6f828d.rna_seq.a...   \n",
       "55  c9432a6e-4d79-4de2-8ffa-e3b05129f78f.rna_seq.a...   \n",
       "19  1432142c-908d-46d7-9a2a-2597a0f00ae8.rna_seq.a...   \n",
       "27  8e595756-ee6b-4fd9-9844-6001ea97fa2e.rna_seq.a...   \n",
       "5   ebbdae1d-a199-4577-a73a-77e9c6a55a92.rna_seq.a...   \n",
       "\n",
       "                                      id  \n",
       "85  0444bb3d-0844-4333-97b0-e71faca95abb  \n",
       "55  0940bd97-2818-48f6-a0a5-f955152409b8  \n",
       "19  095b68ff-e317-4667-b1d1-ff69254d146a  \n",
       "27  0c11ccb8-4405-4a2a-9180-b483fb845968  \n",
       "5   0dbea4cc-32b6-42af-b4f0-fa0a67b05a7f  "
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sorted_id_indices = df_NT.sort_values(by=\"id\").index\n",
    "df_NT.sort_values(by=\"id\").head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6b745807-9da9-486d-b27c-3a5b7e08f9fc",
   "metadata": {},
   "source": [
    "# Download the files\n",
    "\n",
    "We proceed by downloading the gene expression data for our selected samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "e19f6ff0-071d-49a9-a7c3-ab469cb084e7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "File already exists: True\n",
      "Storing temporary files in ../data/liver/tmp\n"
     ]
    }
   ],
   "source": [
    "file_already_exists = os.path.isfile(outfile)\n",
    "print(\"File already exists:\", file_already_exists)\n",
    "\n",
    "tmp = Path(outfolder, \"tmp\")\n",
    "!mkdir -p $tmp\n",
    "print(f\"Storing temporary files in {tmp}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "ffba0ee5-3454-4867-ae81-d26342681120",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>filename</th>\n",
       "      <th>md5</th>\n",
       "      <th>size</th>\n",
       "      <th>state</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>df736f4c-b9d6-4805-bd7d-7bc6d92174d7</td>\n",
       "      <td>df736f4c-b9d6-4805-bd7d-7bc6d92174d7/e576f331-...</td>\n",
       "      <td>8e0c59d0d0efd855bdd59fdd4e84e2d0</td>\n",
       "      <td>4223797</td>\n",
       "      <td>validated</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>6e812349-35e7-44e7-ac7c-a4bc19d81fd0</td>\n",
       "      <td>6e812349-35e7-44e7-ac7c-a4bc19d81fd0/14fc153e-...</td>\n",
       "      <td>6eef53bb76ba2ab3fc3842fb29cfae88</td>\n",
       "      <td>4201688</td>\n",
       "      <td>validated</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>d80985cc-be01-452c-931c-65fccbe783d9</td>\n",
       "      <td>d80985cc-be01-452c-931c-65fccbe783d9/01a67ac4-...</td>\n",
       "      <td>65cd2791cb00e320933c2511b66dd7cf</td>\n",
       "      <td>4220262</td>\n",
       "      <td>validated</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>d139256a-c9e8-417b-90af-c466a58d459d</td>\n",
       "      <td>d139256a-c9e8-417b-90af-c466a58d459d/c0b92413-...</td>\n",
       "      <td>8e93ddc7538d8b7aaa829d040c7e4d25</td>\n",
       "      <td>4209435</td>\n",
       "      <td>validated</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>35c358f0-4afc-4bd3-a0f4-2741134d6f35</td>\n",
       "      <td>35c358f0-4afc-4bd3-a0f4-2741134d6f35/533c5b0b-...</td>\n",
       "      <td>69bc0158cc1444d14c2007cf219b12b2</td>\n",
       "      <td>4227337</td>\n",
       "      <td>validated</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                     id  \\\n",
       "0  df736f4c-b9d6-4805-bd7d-7bc6d92174d7   \n",
       "1  6e812349-35e7-44e7-ac7c-a4bc19d81fd0   \n",
       "2  d80985cc-be01-452c-931c-65fccbe783d9   \n",
       "3  d139256a-c9e8-417b-90af-c466a58d459d   \n",
       "4  35c358f0-4afc-4bd3-a0f4-2741134d6f35   \n",
       "\n",
       "                                            filename  \\\n",
       "0  df736f4c-b9d6-4805-bd7d-7bc6d92174d7/e576f331-...   \n",
       "1  6e812349-35e7-44e7-ac7c-a4bc19d81fd0/14fc153e-...   \n",
       "2  d80985cc-be01-452c-931c-65fccbe783d9/01a67ac4-...   \n",
       "3  d139256a-c9e8-417b-90af-c466a58d459d/c0b92413-...   \n",
       "4  35c358f0-4afc-4bd3-a0f4-2741134d6f35/533c5b0b-...   \n",
       "\n",
       "                                md5     size      state  \n",
       "0  8e0c59d0d0efd855bdd59fdd4e84e2d0  4223797  validated  \n",
       "1  6eef53bb76ba2ab3fc3842fb29cfae88  4201688  validated  \n",
       "2  65cd2791cb00e320933c2511b66dd7cf  4220262  validated  \n",
       "3  8e93ddc7538d8b7aaa829d040c7e4d25  4209435  validated  \n",
       "4  69bc0158cc1444d14c2007cf219b12b2  4227337  validated  "
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "if overwrite or not file_already_exists:\n",
    "\n",
    "    # Download the data\n",
    "    data_endpt = \"https://api.gdc.cancer.gov/data\"\n",
    "    params = {\"ids\": ids}\n",
    "\n",
    "    response = requests.post(data_endpt,\n",
    "                          data = json.dumps(params),\n",
    "                          headers={\n",
    "                              \"Content-Type\": \"application/json\"\n",
    "                              })\n",
    "\n",
    "    response_head_cd = response.headers[\"Content-Disposition\"]\n",
    "\n",
    "    filename = Path(tmp, re.findall(\"filename=(.+)\", response_head_cd)[0])\n",
    "\n",
    "    with open(filename, \"wb\") as output_file:\n",
    "        output_file.write(response.content)\n",
    "\n",
    "    # Unpack the data\n",
    "    !tar -xzf $filename -C $tmp\n",
    "\n",
    "# Sort manifest to get matching NT pairs like in df_NT\n",
    "try:\n",
    "    man = pd.read_csv(Path(tmp, \"MANIFEST.txt\"),sep=\"\\t\")\n",
    "except FileNotFoundError:\n",
    "    man = pd.read_csv(Path(outfolder, \"MANIFEST.txt\"),sep=\"\\t\")\n",
    "\n",
    "man = man.sort_values(by=\"id\")\n",
    "man.index = sorted_id_indices\n",
    "man = man.sort_index()\n",
    "man.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "940e3e16-5b28-4c19-9543-d8497cc77b6a",
   "metadata": {},
   "source": [
    "# Formatting and cleanup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "a515f234-d992-43ac-817f-69c0e95484dd",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0 files processed\n",
      "20 files processed\n",
      "40 files processed\n",
      "60 files processed\n",
      "80 files processed\n"
     ]
    }
   ],
   "source": [
    "if overwrite or not file_already_exists:\n",
    "\n",
    "    df_tmp = pd.DataFrame()\n",
    "\n",
    "    for i, filename in  enumerate(man[\"filename\"]):\n",
    "\n",
    "        filename = Path(tmp, filename)\n",
    "        \n",
    "        # if Path(filename).is_file(): \n",
    "        #     !gzip -d $filename\n",
    "    \n",
    "        f = pd.read_csv(filename, sep=\"\\t\",header=[1], index_col=0)\n",
    "\n",
    "        if i%2==0: \n",
    "            sample_type = \"T\"\n",
    "            sample_number = i//2 + 1\n",
    "        else:\n",
    "            sample_type = \"N\"\n",
    "            sample_number = (i+1)//2\n",
    "\n",
    "        df_tmp[f\"{sample_type}{sample_number}\"] = f[\"unstranded\"]\n",
    "        \n",
    "        if i % 20 == 0: print(i, \"files processed\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 133,
   "id": "9ee805d2-863d-4673-bd27-888ab8cbd3fd",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loading existing file: ../data/liver/liver.csv\n"
     ]
    }
   ],
   "source": [
    "rows_to_remove = [\"N_unmapped\",\"N_multimapping\",\"N_noFeature\",\"N_ambiguous\"]\n",
    "\n",
    "if overwrite or not file_already_exists:\n",
    "    df_out = df_tmp.drop(rows_to_remove, axis=0)\n",
    "else:\n",
    "    print(\"Loading existing file:\", outfile)\n",
    "    df_out = pd.read_csv(outfile, index_col = 0)\n",
    "    df_meta = pd.read_csv(meta_outfile, index_col=0)\n",
    "\n",
    "replicates = len(df_out.columns)//2\n",
    "cols = [f\"N{i}\" for i in range(1,replicates+1)] + [f\"T{i}\" for i in range(1,replicates+1)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "bd24f55a-42f8-44b1-81db-b4f645a3a0eb",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(16, 100)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>N1</th>\n",
       "      <th>N2</th>\n",
       "      <th>N3</th>\n",
       "      <th>N4</th>\n",
       "      <th>N5</th>\n",
       "      <th>N6</th>\n",
       "      <th>N7</th>\n",
       "      <th>N8</th>\n",
       "      <th>N9</th>\n",
       "      <th>N10</th>\n",
       "      <th>...</th>\n",
       "      <th>T41</th>\n",
       "      <th>T42</th>\n",
       "      <th>T43</th>\n",
       "      <th>T44</th>\n",
       "      <th>T45</th>\n",
       "      <th>T46</th>\n",
       "      <th>T47</th>\n",
       "      <th>T48</th>\n",
       "      <th>T49</th>\n",
       "      <th>T50</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>days_to_birth</th>\n",
       "      <td>-26400.0</td>\n",
       "      <td>-22968.0</td>\n",
       "      <td>-25489.0</td>\n",
       "      <td>-27756.0</td>\n",
       "      <td>-20188.0</td>\n",
       "      <td>-15410.0</td>\n",
       "      <td>-24934.0</td>\n",
       "      <td>-28406.0</td>\n",
       "      <td>-14613.0</td>\n",
       "      <td>-24646.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-28101.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-27530.0</td>\n",
       "      <td>-18562.0</td>\n",
       "      <td>-19473.0</td>\n",
       "      <td>-19140.0</td>\n",
       "      <td>-28049.0</td>\n",
       "      <td>-22913.0</td>\n",
       "      <td>-18951.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>days_to_death</th>\n",
       "      <td>1135.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1149.0</td>\n",
       "      <td>1622.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>223.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>581.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>247.0</td>\n",
       "      <td>194.0</td>\n",
       "      <td>91.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>770.0</td>\n",
       "      <td>711.0</td>\n",
       "      <td>34.0</td>\n",
       "      <td>2116.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ethnicity</th>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>not reported</td>\n",
       "      <td>not reported</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>...</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gender</th>\n",
       "      <td>female</td>\n",
       "      <td>female</td>\n",
       "      <td>male</td>\n",
       "      <td>female</td>\n",
       "      <td>female</td>\n",
       "      <td>male</td>\n",
       "      <td>male</td>\n",
       "      <td>female</td>\n",
       "      <td>male</td>\n",
       "      <td>male</td>\n",
       "      <td>...</td>\n",
       "      <td>male</td>\n",
       "      <td>female</td>\n",
       "      <td>female</td>\n",
       "      <td>male</td>\n",
       "      <td>male</td>\n",
       "      <td>male</td>\n",
       "      <td>female</td>\n",
       "      <td>male</td>\n",
       "      <td>female</td>\n",
       "      <td>female</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>race</th>\n",
       "      <td>white</td>\n",
       "      <td>white</td>\n",
       "      <td>white</td>\n",
       "      <td>black or african american</td>\n",
       "      <td>white</td>\n",
       "      <td>black or african american</td>\n",
       "      <td>asian</td>\n",
       "      <td>white</td>\n",
       "      <td>asian</td>\n",
       "      <td>black or african american</td>\n",
       "      <td>...</td>\n",
       "      <td>white</td>\n",
       "      <td>white</td>\n",
       "      <td>white</td>\n",
       "      <td>white</td>\n",
       "      <td>asian</td>\n",
       "      <td>asian</td>\n",
       "      <td>white</td>\n",
       "      <td>white</td>\n",
       "      <td>white</td>\n",
       "      <td>black or african american</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 100 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                   N1            N2            N3  \\\n",
       "days_to_birth                -26400.0      -22968.0      -25489.0   \n",
       "days_to_death                  1135.0           NaN           NaN   \n",
       "ethnicity      not hispanic or latino  not reported  not reported   \n",
       "gender                         female        female          male   \n",
       "race                            white         white         white   \n",
       "\n",
       "                                      N4                      N5  \\\n",
       "days_to_birth                   -27756.0                -20188.0   \n",
       "days_to_death                        NaN                     NaN   \n",
       "ethnicity         not hispanic or latino  not hispanic or latino   \n",
       "gender                            female                  female   \n",
       "race           black or african american                   white   \n",
       "\n",
       "                                      N6                      N7  \\\n",
       "days_to_birth                   -15410.0                -24934.0   \n",
       "days_to_death                     1149.0                  1622.0   \n",
       "ethnicity         not hispanic or latino  not hispanic or latino   \n",
       "gender                              male                    male   \n",
       "race           black or african american                   asian   \n",
       "\n",
       "                                   N8                      N9  \\\n",
       "days_to_birth                -28406.0                -14613.0   \n",
       "days_to_death                     NaN                   223.0   \n",
       "ethnicity      not hispanic or latino  not hispanic or latino   \n",
       "gender                         female                    male   \n",
       "race                            white                   asian   \n",
       "\n",
       "                                     N10  ...                     T41  \\\n",
       "days_to_birth                   -24646.0  ...                     NaN   \n",
       "days_to_death                        NaN  ...                   581.0   \n",
       "ethnicity         not hispanic or latino  ...  not hispanic or latino   \n",
       "gender                              male  ...                    male   \n",
       "race           black or african american  ...                   white   \n",
       "\n",
       "                                  T42                     T43  \\\n",
       "days_to_birth                -28101.0                     NaN   \n",
       "days_to_death                     NaN                   247.0   \n",
       "ethnicity      not hispanic or latino  not hispanic or latino   \n",
       "gender                         female                  female   \n",
       "race                            white                   white   \n",
       "\n",
       "                                  T44                     T45  \\\n",
       "days_to_birth                -27530.0                -18562.0   \n",
       "days_to_death                   194.0                    91.0   \n",
       "ethnicity      not hispanic or latino  not hispanic or latino   \n",
       "gender                           male                    male   \n",
       "race                            white                   asian   \n",
       "\n",
       "                                  T46                     T47  \\\n",
       "days_to_birth                -19473.0                -19140.0   \n",
       "days_to_death                     NaN                   770.0   \n",
       "ethnicity      not hispanic or latino  not hispanic or latino   \n",
       "gender                           male                  female   \n",
       "race                            asian                   white   \n",
       "\n",
       "                                  T48                     T49  \\\n",
       "days_to_birth                -28049.0                -22913.0   \n",
       "days_to_death                   711.0                    34.0   \n",
       "ethnicity      not hispanic or latino  not hispanic or latino   \n",
       "gender                           male                  female   \n",
       "race                            white                   white   \n",
       "\n",
       "                                     T50  \n",
       "days_to_birth                   -18951.0  \n",
       "days_to_death                     2116.0  \n",
       "ethnicity         not hispanic or latino  \n",
       "gender                            female  \n",
       "race           black or african american  \n",
       "\n",
       "[5 rows x 100 columns]"
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "NT_cols = [f'T{i//2 + 1}' if i%2==0 else f'N{(i+1)//2}' for i in range(len(man))] # sorted like df_NT\n",
    "sample_metadata = df_NT\n",
    "sample_metadata[\"case\"] = [c[0] for c in NT_cols]\n",
    "sample_metadata.index = NT_cols\n",
    "sample_metadata = sample_metadata.sort_index()\n",
    "sample_metadata.index = cols\n",
    "\n",
    "assert(np.all(sample_metadata[:replicates][\"submitter_id\"].values == sample_metadata[replicates:][\"submitter_id\"].values))\n",
    "assert(np.all(sample_metadata[:replicates][\"disease_type\"].values == sample_metadata[replicates:][\"disease_type\"].values))\n",
    "sample_metadata = sample_metadata.T\n",
    "print(sample_metadata.shape)\n",
    "sample_metadata.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "id": "1c485119-e5e1-4c45-b14d-3c07e84520b8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "44 duplicate indices dropped\n",
      "(60616, 100)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>N1</th>\n",
       "      <th>N2</th>\n",
       "      <th>N3</th>\n",
       "      <th>N4</th>\n",
       "      <th>N5</th>\n",
       "      <th>N6</th>\n",
       "      <th>N7</th>\n",
       "      <th>N8</th>\n",
       "      <th>N9</th>\n",
       "      <th>N10</th>\n",
       "      <th>...</th>\n",
       "      <th>T41</th>\n",
       "      <th>T42</th>\n",
       "      <th>T43</th>\n",
       "      <th>T44</th>\n",
       "      <th>T45</th>\n",
       "      <th>T46</th>\n",
       "      <th>T47</th>\n",
       "      <th>T48</th>\n",
       "      <th>T49</th>\n",
       "      <th>T50</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000000003</th>\n",
       "      <td>4303</td>\n",
       "      <td>4847</td>\n",
       "      <td>4199</td>\n",
       "      <td>3059</td>\n",
       "      <td>3621</td>\n",
       "      <td>5268</td>\n",
       "      <td>5224</td>\n",
       "      <td>3008</td>\n",
       "      <td>4803</td>\n",
       "      <td>5692</td>\n",
       "      <td>...</td>\n",
       "      <td>8836</td>\n",
       "      <td>16693</td>\n",
       "      <td>7225</td>\n",
       "      <td>5569</td>\n",
       "      <td>4008</td>\n",
       "      <td>3752</td>\n",
       "      <td>5699</td>\n",
       "      <td>3954</td>\n",
       "      <td>5665</td>\n",
       "      <td>2321</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000005</th>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>7</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>16</td>\n",
       "      <td>3</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>22</td>\n",
       "      <td>12</td>\n",
       "      <td>10</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000419</th>\n",
       "      <td>802</td>\n",
       "      <td>964</td>\n",
       "      <td>1063</td>\n",
       "      <td>703</td>\n",
       "      <td>657</td>\n",
       "      <td>1399</td>\n",
       "      <td>1093</td>\n",
       "      <td>777</td>\n",
       "      <td>795</td>\n",
       "      <td>1043</td>\n",
       "      <td>...</td>\n",
       "      <td>1454</td>\n",
       "      <td>1419</td>\n",
       "      <td>866</td>\n",
       "      <td>2055</td>\n",
       "      <td>2124</td>\n",
       "      <td>909</td>\n",
       "      <td>1295</td>\n",
       "      <td>2241</td>\n",
       "      <td>1920</td>\n",
       "      <td>542</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000457</th>\n",
       "      <td>411</td>\n",
       "      <td>263</td>\n",
       "      <td>485</td>\n",
       "      <td>256</td>\n",
       "      <td>396</td>\n",
       "      <td>577</td>\n",
       "      <td>479</td>\n",
       "      <td>279</td>\n",
       "      <td>353</td>\n",
       "      <td>600</td>\n",
       "      <td>...</td>\n",
       "      <td>777</td>\n",
       "      <td>543</td>\n",
       "      <td>673</td>\n",
       "      <td>620</td>\n",
       "      <td>2940</td>\n",
       "      <td>605</td>\n",
       "      <td>585</td>\n",
       "      <td>859</td>\n",
       "      <td>642</td>\n",
       "      <td>184</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000460</th>\n",
       "      <td>69</td>\n",
       "      <td>113</td>\n",
       "      <td>100</td>\n",
       "      <td>78</td>\n",
       "      <td>131</td>\n",
       "      <td>95</td>\n",
       "      <td>77</td>\n",
       "      <td>82</td>\n",
       "      <td>303</td>\n",
       "      <td>285</td>\n",
       "      <td>...</td>\n",
       "      <td>367</td>\n",
       "      <td>132</td>\n",
       "      <td>319</td>\n",
       "      <td>219</td>\n",
       "      <td>1387</td>\n",
       "      <td>108</td>\n",
       "      <td>236</td>\n",
       "      <td>704</td>\n",
       "      <td>359</td>\n",
       "      <td>42</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 100 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   N1    N2    N3    N4    N5    N6    N7    N8    N9   N10  \\\n",
       "                                                                              \n",
       "ENSG00000000003  4303  4847  4199  3059  3621  5268  5224  3008  4803  5692   \n",
       "ENSG00000000005     1     3     4     0     0     7     4     1    16     3   \n",
       "ENSG00000000419   802   964  1063   703   657  1399  1093   777   795  1043   \n",
       "ENSG00000000457   411   263   485   256   396   577   479   279   353   600   \n",
       "ENSG00000000460    69   113   100    78   131    95    77    82   303   285   \n",
       "\n",
       "                 ...   T41    T42   T43   T44   T45   T46   T47   T48   T49  \\\n",
       "                 ...                                                          \n",
       "ENSG00000000003  ...  8836  16693  7225  5569  4008  3752  5699  3954  5665   \n",
       "ENSG00000000005  ...     0      0     0     1     1     4    22    12    10   \n",
       "ENSG00000000419  ...  1454   1419   866  2055  2124   909  1295  2241  1920   \n",
       "ENSG00000000457  ...   777    543   673   620  2940   605   585   859   642   \n",
       "ENSG00000000460  ...   367    132   319   219  1387   108   236   704   359   \n",
       "\n",
       "                  T50  \n",
       "                       \n",
       "ENSG00000000003  2321  \n",
       "ENSG00000000005     2  \n",
       "ENSG00000000419   542  \n",
       "ENSG00000000457   184  \n",
       "ENSG00000000460    42  \n",
       "\n",
       "[5 rows x 100 columns]"
      ]
     },
     "execution_count": 99,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "if \".\" in df_out.index[0]:\n",
    "    df_out = df_out.sort_index(axis=1)\n",
    "    df_out.index = df_out.index.map(lambda x: x.split(\".\")[0])\n",
    "    df_out.index.name = \"\"\n",
    "    df_out.columns = cols\n",
    "    \n",
    "dupes = df_out.index.duplicated(keep=\"first\")\n",
    "print(f\"{dupes.sum()} duplicate indices dropped\")\n",
    "df_out = df_out[~dupes]\n",
    "\n",
    "print(df_out.shape)\n",
    "df_out.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "id": "6e8796e3-061d-4604-91d3-0e5607f15bb7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Metadata saved in ../data/liver/liver_meta.csv\n",
      "Full data saved in ../data/liver/liver.csv\n"
     ]
    }
   ],
   "source": [
    "if overwrite or not file_already_exists:\n",
    "    \n",
    "    meta_outfile = Path(outfile._str[:-4]+\"_meta.csv\")\n",
    "    sample_metadata.to_csv(meta_outfile)\n",
    "    print(f\"Metadata saved in {meta_outfile}\")\n",
    "\n",
    "    # Save full count data (no genes filtered)\n",
    "    outfile_full = Path(outfolder, f\"{primary_site.lower()}.csv\")\n",
    "    df_out.to_csv(outfile_full)\n",
    "    print(f\"Full data saved in {outfile_full}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "80bd3ec5-6e92-4d21-8145-fe435979476b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Cleaning up tmp folder\n"
     ]
    }
   ],
   "source": [
    "if cleanup: \n",
    "    print(\"Cleaning up tmp folder\")\n",
    "    !mv $tmp/MANIFEST.txt $tmp/..\n",
    "    !rm -r $tmp"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "467ba9d8-6f7b-4ae4-9f1c-00567eb18f28",
   "metadata": {},
   "source": [
    "# Filter by expression\n",
    "\n",
    "Use edgeR's filterByExpr() function to remove low count genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 121,
   "id": "82596d3a-9d6d-4356-92bf-b4de2037451e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "PosixPath('../data/liver/liver_filtered.csv')"
      ]
     },
     "execution_count": 121,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "outfile_filtered"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 119,
   "id": "2429888f-f1fd-482e-aecf-1a930bc383cb",
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys, importlib\n",
    "importlib.reload(sys.modules['R_wrappers'])\n",
    "\n",
    "modpath = Path(\"../scripts\")\n",
    "sys.path.append(os.path.abspath(modpath))\n",
    "\n",
    "from R_wrappers import filterByExpr_wrapper\n",
    "\n",
    "outfile_filtered = Path(outfile._str.replace(\".csv\",\"_filtered.csv\"))\n",
    "filterByExpr_wrapper(inpath=outfile._str, outpath=outfile_filtered, design=\"none\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 162,
   "id": "93607aeb-40a4-4258-b73c-6e8eea209435",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(15263, 100)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>N1</th>\n",
       "      <th>N2</th>\n",
       "      <th>N3</th>\n",
       "      <th>N4</th>\n",
       "      <th>N5</th>\n",
       "      <th>N6</th>\n",
       "      <th>N7</th>\n",
       "      <th>N8</th>\n",
       "      <th>N9</th>\n",
       "      <th>N10</th>\n",
       "      <th>...</th>\n",
       "      <th>T41</th>\n",
       "      <th>T42</th>\n",
       "      <th>T43</th>\n",
       "      <th>T44</th>\n",
       "      <th>T45</th>\n",
       "      <th>T46</th>\n",
       "      <th>T47</th>\n",
       "      <th>T48</th>\n",
       "      <th>T49</th>\n",
       "      <th>T50</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000000003</th>\n",
       "      <td>4303</td>\n",
       "      <td>4847</td>\n",
       "      <td>4199</td>\n",
       "      <td>3059</td>\n",
       "      <td>3621</td>\n",
       "      <td>5268</td>\n",
       "      <td>5224</td>\n",
       "      <td>3008</td>\n",
       "      <td>4803</td>\n",
       "      <td>5692</td>\n",
       "      <td>...</td>\n",
       "      <td>8836</td>\n",
       "      <td>16693</td>\n",
       "      <td>7225</td>\n",
       "      <td>5569</td>\n",
       "      <td>4008</td>\n",
       "      <td>3752</td>\n",
       "      <td>5699</td>\n",
       "      <td>3954</td>\n",
       "      <td>5665</td>\n",
       "      <td>2321</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000419</th>\n",
       "      <td>802</td>\n",
       "      <td>964</td>\n",
       "      <td>1063</td>\n",
       "      <td>703</td>\n",
       "      <td>657</td>\n",
       "      <td>1399</td>\n",
       "      <td>1093</td>\n",
       "      <td>777</td>\n",
       "      <td>795</td>\n",
       "      <td>1043</td>\n",
       "      <td>...</td>\n",
       "      <td>1454</td>\n",
       "      <td>1419</td>\n",
       "      <td>866</td>\n",
       "      <td>2055</td>\n",
       "      <td>2124</td>\n",
       "      <td>909</td>\n",
       "      <td>1295</td>\n",
       "      <td>2241</td>\n",
       "      <td>1920</td>\n",
       "      <td>542</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000457</th>\n",
       "      <td>411</td>\n",
       "      <td>263</td>\n",
       "      <td>485</td>\n",
       "      <td>256</td>\n",
       "      <td>396</td>\n",
       "      <td>577</td>\n",
       "      <td>479</td>\n",
       "      <td>279</td>\n",
       "      <td>353</td>\n",
       "      <td>600</td>\n",
       "      <td>...</td>\n",
       "      <td>777</td>\n",
       "      <td>543</td>\n",
       "      <td>673</td>\n",
       "      <td>620</td>\n",
       "      <td>2940</td>\n",
       "      <td>605</td>\n",
       "      <td>585</td>\n",
       "      <td>859</td>\n",
       "      <td>642</td>\n",
       "      <td>184</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000460</th>\n",
       "      <td>69</td>\n",
       "      <td>113</td>\n",
       "      <td>100</td>\n",
       "      <td>78</td>\n",
       "      <td>131</td>\n",
       "      <td>95</td>\n",
       "      <td>77</td>\n",
       "      <td>82</td>\n",
       "      <td>303</td>\n",
       "      <td>285</td>\n",
       "      <td>...</td>\n",
       "      <td>367</td>\n",
       "      <td>132</td>\n",
       "      <td>319</td>\n",
       "      <td>219</td>\n",
       "      <td>1387</td>\n",
       "      <td>108</td>\n",
       "      <td>236</td>\n",
       "      <td>704</td>\n",
       "      <td>359</td>\n",
       "      <td>42</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000938</th>\n",
       "      <td>293</td>\n",
       "      <td>231</td>\n",
       "      <td>1476</td>\n",
       "      <td>362</td>\n",
       "      <td>128</td>\n",
       "      <td>630</td>\n",
       "      <td>285</td>\n",
       "      <td>195</td>\n",
       "      <td>209</td>\n",
       "      <td>596</td>\n",
       "      <td>...</td>\n",
       "      <td>214</td>\n",
       "      <td>188</td>\n",
       "      <td>366</td>\n",
       "      <td>96</td>\n",
       "      <td>204</td>\n",
       "      <td>290</td>\n",
       "      <td>327</td>\n",
       "      <td>193</td>\n",
       "      <td>68</td>\n",
       "      <td>333</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 100 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   N1    N2    N3    N4    N5    N6    N7    N8    N9   N10  \\\n",
       "ENSG00000000003  4303  4847  4199  3059  3621  5268  5224  3008  4803  5692   \n",
       "ENSG00000000419   802   964  1063   703   657  1399  1093   777   795  1043   \n",
       "ENSG00000000457   411   263   485   256   396   577   479   279   353   600   \n",
       "ENSG00000000460    69   113   100    78   131    95    77    82   303   285   \n",
       "ENSG00000000938   293   231  1476   362   128   630   285   195   209   596   \n",
       "\n",
       "                 ...   T41    T42   T43   T44   T45   T46   T47   T48   T49  \\\n",
       "ENSG00000000003  ...  8836  16693  7225  5569  4008  3752  5699  3954  5665   \n",
       "ENSG00000000419  ...  1454   1419   866  2055  2124   909  1295  2241  1920   \n",
       "ENSG00000000457  ...   777    543   673   620  2940   605   585   859   642   \n",
       "ENSG00000000460  ...   367    132   319   219  1387   108   236   704   359   \n",
       "ENSG00000000938  ...   214    188   366    96   204   290   327   193    68   \n",
       "\n",
       "                  T50  \n",
       "ENSG00000000003  2321  \n",
       "ENSG00000000419   542  \n",
       "ENSG00000000457   184  \n",
       "ENSG00000000460    42  \n",
       "ENSG00000000938   333  \n",
       "\n",
       "[5 rows x 100 columns]"
      ]
     },
     "execution_count": 162,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_filt = pd.read_csv(outfile_filtered, index_col = 0)\n",
    "print(df_filt.shape)\n",
    "df_filt.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8a1c1e97-877c-438e-8b71-d2804df8d5f8",
   "metadata": {},
   "source": [
    "# Split by project_id and disease_type"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 151,
   "id": "046df570-d92d-4d7a-83d4-d2a46549ade4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "TCGA-LIHC: 100 samples\n",
      "Adenomas and Adenocarcinomas    100\n",
      "Name: disease_type, dtype: int64 \n",
      "\n"
     ]
    }
   ],
   "source": [
    "df = pd.read_csv(outfile_filtered, index_col=0, header=0)\n",
    "df_meta = pd.read_csv(meta_outfile, index_col=0, header=0)\n",
    "\n",
    "for project_id in set(df_meta.loc[\"project_id\"]):\n",
    "    mask = (df_meta.loc[\"project_id\"] == project_id)\n",
    "    mask = mask[mask>0].index\n",
    "    df_p = df_meta.loc[:,mask]\n",
    "    print(f\"{project_id}: {len(df_p.columns)} samples\")\n",
    "    print(df_p.loc[\"disease_type\"].value_counts(),\"\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 152,
   "id": "b64bcf13-3b41-41a0-a852-9c5a34559151",
   "metadata": {},
   "outputs": [],
   "source": [
    "from misc import add_metadata_to_multiindex\n",
    "\n",
    "df_merged = add_metadata_to_multiindex(df,df_meta)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 153,
   "id": "15eca9aa-36ee-487c-8d27-5579f361e7bb",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Data saved in ../data/liver/LIHC/LIHC.csv\n"
     ]
    }
   ],
   "source": [
    "project_ids = set(df_merged.columns.get_level_values(level=\"project_id\"))\n",
    "\n",
    "for project_id in project_ids:\n",
    "    \n",
    "    if project_id in [\"TCGA-KIRC\", \"TCGA-KIRP\", \"TCGA-KICH\", \"TCGA-LUAD\", \n",
    "                      \"TCGA-THCA\", \"TCGA-PRAD\", \"TCGA-COAD\", \"TCGA-LIHC\"]:\n",
    "        disease_type = \"Adenomas and Adenocarcinomas\"\n",
    "    elif project_id in [\"TCGA-BRCA\"]:\n",
    "        disease_type = \"Ductal and Lobular Neoplasms\"\n",
    "    elif project_id in [\"TCGA-LUSC\"]:\n",
    "        disease_type = \"Squamous Cell Neoplasms\"\n",
    "    elif project_id in [\"TCGA-READ\"]:\n",
    "        continue\n",
    "    else: \n",
    "        raise Exception(f\"Specify disease_type for {project_id}\")\n",
    "        \n",
    "    project_id_abbrev = project_id[5:] # drop TCGA- prefix\n",
    "    project_outfolder = Path(outfolder, project_id_abbrev)\n",
    "    os.system(f\"mkdir -p {project_outfolder}\")\n",
    "    \n",
    "    df_xs = df_merged.xs(project_id, level=\"project_id\", axis=1)\n",
    "    df_xs = df_xs.xs(disease_type, level=\"disease_type\", axis=1)\n",
    "    \n",
    "    project_sample_columns = df_xs.columns.get_level_values(level=\"Sample\")\n",
    "    df_project = df[project_sample_columns]\n",
    "    \n",
    "    outfile_project = Path(project_outfolder, project_id_abbrev+\".csv\")\n",
    "    df_project.to_csv(outfile_project)\n",
    "    print(f\"Data saved in {outfile_project}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 157,
   "id": "98530304-0ce3-411a-b0cf-ccc1963f3973",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(15263, 100)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>N1</th>\n",
       "      <th>N2</th>\n",
       "      <th>N3</th>\n",
       "      <th>N4</th>\n",
       "      <th>N5</th>\n",
       "      <th>N6</th>\n",
       "      <th>N7</th>\n",
       "      <th>N8</th>\n",
       "      <th>N9</th>\n",
       "      <th>N10</th>\n",
       "      <th>...</th>\n",
       "      <th>T41</th>\n",
       "      <th>T42</th>\n",
       "      <th>T43</th>\n",
       "      <th>T44</th>\n",
       "      <th>T45</th>\n",
       "      <th>T46</th>\n",
       "      <th>T47</th>\n",
       "      <th>T48</th>\n",
       "      <th>T49</th>\n",
       "      <th>T50</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000000003</th>\n",
       "      <td>4303</td>\n",
       "      <td>4847</td>\n",
       "      <td>4199</td>\n",
       "      <td>3059</td>\n",
       "      <td>3621</td>\n",
       "      <td>5268</td>\n",
       "      <td>5224</td>\n",
       "      <td>3008</td>\n",
       "      <td>4803</td>\n",
       "      <td>5692</td>\n",
       "      <td>...</td>\n",
       "      <td>8836</td>\n",
       "      <td>16693</td>\n",
       "      <td>7225</td>\n",
       "      <td>5569</td>\n",
       "      <td>4008</td>\n",
       "      <td>3752</td>\n",
       "      <td>5699</td>\n",
       "      <td>3954</td>\n",
       "      <td>5665</td>\n",
       "      <td>2321</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000419</th>\n",
       "      <td>802</td>\n",
       "      <td>964</td>\n",
       "      <td>1063</td>\n",
       "      <td>703</td>\n",
       "      <td>657</td>\n",
       "      <td>1399</td>\n",
       "      <td>1093</td>\n",
       "      <td>777</td>\n",
       "      <td>795</td>\n",
       "      <td>1043</td>\n",
       "      <td>...</td>\n",
       "      <td>1454</td>\n",
       "      <td>1419</td>\n",
       "      <td>866</td>\n",
       "      <td>2055</td>\n",
       "      <td>2124</td>\n",
       "      <td>909</td>\n",
       "      <td>1295</td>\n",
       "      <td>2241</td>\n",
       "      <td>1920</td>\n",
       "      <td>542</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000457</th>\n",
       "      <td>411</td>\n",
       "      <td>263</td>\n",
       "      <td>485</td>\n",
       "      <td>256</td>\n",
       "      <td>396</td>\n",
       "      <td>577</td>\n",
       "      <td>479</td>\n",
       "      <td>279</td>\n",
       "      <td>353</td>\n",
       "      <td>600</td>\n",
       "      <td>...</td>\n",
       "      <td>777</td>\n",
       "      <td>543</td>\n",
       "      <td>673</td>\n",
       "      <td>620</td>\n",
       "      <td>2940</td>\n",
       "      <td>605</td>\n",
       "      <td>585</td>\n",
       "      <td>859</td>\n",
       "      <td>642</td>\n",
       "      <td>184</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000460</th>\n",
       "      <td>69</td>\n",
       "      <td>113</td>\n",
       "      <td>100</td>\n",
       "      <td>78</td>\n",
       "      <td>131</td>\n",
       "      <td>95</td>\n",
       "      <td>77</td>\n",
       "      <td>82</td>\n",
       "      <td>303</td>\n",
       "      <td>285</td>\n",
       "      <td>...</td>\n",
       "      <td>367</td>\n",
       "      <td>132</td>\n",
       "      <td>319</td>\n",
       "      <td>219</td>\n",
       "      <td>1387</td>\n",
       "      <td>108</td>\n",
       "      <td>236</td>\n",
       "      <td>704</td>\n",
       "      <td>359</td>\n",
       "      <td>42</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000938</th>\n",
       "      <td>293</td>\n",
       "      <td>231</td>\n",
       "      <td>1476</td>\n",
       "      <td>362</td>\n",
       "      <td>128</td>\n",
       "      <td>630</td>\n",
       "      <td>285</td>\n",
       "      <td>195</td>\n",
       "      <td>209</td>\n",
       "      <td>596</td>\n",
       "      <td>...</td>\n",
       "      <td>214</td>\n",
       "      <td>188</td>\n",
       "      <td>366</td>\n",
       "      <td>96</td>\n",
       "      <td>204</td>\n",
       "      <td>290</td>\n",
       "      <td>327</td>\n",
       "      <td>193</td>\n",
       "      <td>68</td>\n",
       "      <td>333</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 100 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   N1    N2    N3    N4    N5    N6    N7    N8    N9   N10  \\\n",
       "ENSG00000000003  4303  4847  4199  3059  3621  5268  5224  3008  4803  5692   \n",
       "ENSG00000000419   802   964  1063   703   657  1399  1093   777   795  1043   \n",
       "ENSG00000000457   411   263   485   256   396   577   479   279   353   600   \n",
       "ENSG00000000460    69   113   100    78   131    95    77    82   303   285   \n",
       "ENSG00000000938   293   231  1476   362   128   630   285   195   209   596   \n",
       "\n",
       "                 ...   T41    T42   T43   T44   T45   T46   T47   T48   T49  \\\n",
       "ENSG00000000003  ...  8836  16693  7225  5569  4008  3752  5699  3954  5665   \n",
       "ENSG00000000419  ...  1454   1419   866  2055  2124   909  1295  2241  1920   \n",
       "ENSG00000000457  ...   777    543   673   620  2940   605   585   859   642   \n",
       "ENSG00000000460  ...   367    132   319   219  1387   108   236   704   359   \n",
       "ENSG00000000938  ...   214    188   366    96   204   290   327   193    68   \n",
       "\n",
       "                  T50  \n",
       "ENSG00000000003  2321  \n",
       "ENSG00000000419   542  \n",
       "ENSG00000000457   184  \n",
       "ENSG00000000460    42  \n",
       "ENSG00000000938   333  \n",
       "\n",
       "[5 rows x 100 columns]"
      ]
     },
     "execution_count": 157,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_project = pd.read_csv(outfile_project, index_col=0)\n",
    "print(df_project.shape)\n",
    "df_project.head()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
